### PR TITLE
Ignoring Task entities with no step key

### PR DESCRIPTION
**Pull request by jakubjezek001**
# Change description
Some productions might be using tasks without SG task entity `step` key. This will change will not raise the missing key and therefore it will allow to sync tasks into project. New settings for defining default task type were added under Compatibility Settigs section. This way we can assign task type to those tasks which are missing pipeline step at SG entity. 

# Testing steps:
1. Make sure `ayon+settings://shotgrid/compatibility_settings/default_task_type` is set at your Project settings with existing Anatomy Task type. Ideally create new one and name it `Unknown`
2. Go to any of yours test project at SG and create new task under any Shot or Asset.
3. Make usre the autosync is set to both direction on SG and AY side.
4. Task shoud be created and the default task type is assigned to it. 
5. Remove the task on AY side and notice that after refreash it had been also retired on SG side.

### CODE REVIEW

**Comment by MilaKudr** [comment]
During manual sync I got these errors:

Traceback (most recent call last):
  File "C:\Users\Mila\code\ayon-shotgrid\services\processor\processor\processor.py", line 243, in start_processing
    handler.process_event(
  File "C:\Users\Mila\code\ayon-shotgrid\services\processor\processor\handlers\sync_projects.py", line 34, in process_event
    hub.synchronize_projects(source=sync_source)
  File "C:\Users\Mila\code\ayon-shotgrid\services\shotgrid_common\ayon_shotgrid_hub\__init__.py", line 277, in synchronize_projects
    match_shotgrid_hierarchy_in_ayon(
  File "C:\Users\Mila\code\ayon-shotgrid\services\shotgrid_common\ayon_shotgrid_hub\match_shotgrid_hierarchy_in_ayon.py", line 118, in match_shotgrid_hierarchy_in_ayon
    ay_entity = _create_new_entity(
  File "C:\Users\Mila\code\ayon-shotgrid\services\shotgrid_common\ayon_shotgrid_hub\match_shotgrid_hierarchy_in_ayon.py", line 259, in _create_new_entity
    sg_ay_dict["task_type"],
KeyError: 'task_type'
2024-09-16 10:25:53.128 | processor | start_processing | INFO: Event has been processed... setting to finished!

**Comment by fabiaserra** [comment]
is this ignoring them or actually syncing them? the title says it's ignoring but the description says:
> This will change will not raise the missing key and therefore it will allow to sync tasks into project

What's the intended behaviour? if you don't set a task type on the Ayon entity what does it happen?

**Comment by jakubjezek001** [comment]
> is this ignoring them or actually syncing them? the title says it's ignoring but the description says:
> 
> > This will change will not raise the missing key and therefore it will allow to sync tasks into project
> 
> What's the intended behaviour? if you don't set a task type on the Ayon entity what does it happen?

Good point, @fabiaserra. It should ignore them and not produce an error message. We had a case where a client was using SG for development ticket submission. These tasks were in the CG project but didn't have any steps (task types).

  

There are two solutions:

  

1. Ignore them so they aren't synchronized into AYON at all.
2. Assign them a default task type that AYON offers.
3. Offer a settings toggle to choose between the two options mentioned above.

  

@fabiaserra or @robin-ynput, do you have an opinion on this?

**Comment by fabiaserra** [comment]
> There are two solutions:
> 
> 1. Ignore them so they aren't synchronized into AYON at all.
> 2. Assign them a default task type that AYON offers.
> 3. Offer a settings toggle to choose between the two options mentioned above.
> 
> @fabiaserra or @robin-ynput, do you have an opinion on this?

Having the settings control is nice but I think most people are confused with the amount of settings control and it kind of beats the purpose. I think I would probably expect them to sync and use the task name for type in order to sync from SG to AYON and for AYON to SG if there's no step that exists for that type just not set the step when creating the task

**Comment by robin-ynput** [comment]
> > is this ignoring them or actually syncing them? the title says it's ignoring but the description says:
> > > This will change will not raise the missing key and therefore it will allow to sync tasks into project
> > 
> > 
> > What's the intended behaviour? if you don't set a task type on the Ayon entity what does it happen?
> 
> Good point, @fabiaserra. It should ignore them and not produce an error message. We had a case where a client was using SG for development ticket submission. These tasks were in the CG project but didn't have any steps (task types).
> 
> There are two solutions:
> 
> 1. Ignore them so they aren't synchronized into AYON at all.
> 2. Assign them a default task type that AYON offers.
> 3. Offer a settings toggle to choose between the two options mentioned above.
> 
> @fabiaserra or @robin-ynput, do you have an opinion on this?

My understanding is that while `Pipeline Step` are optional in SG, `Task Type` are mandatory in AYON.
So I'd recommend to always sync tasks but either set them as `Generic` or a new `Unknown` type in AYON.

**Comment by mkolar** [comment]
Yeah, if pipeline steps are optional in SG, we need to make a reasonable fallback choice. I like the `unknown` type actually.

**Comment by jakubjezek001** [comment]
> Yeah, if pipeline steps are optional in SG, we need to make a reasonable fallback choice. I like the `unknown` type actually.

Here is the solution added to the current state of this PR. Settings are having new option under Compatibility Settings section for defining a default task type. 
image

**Review by MilaKudr — APPROVED** [review]
Works well, tasks are created without any errors on the server.
Screenshot 2024-12-04 095322
Screenshot 2024-12-04 095413

**Review by MilaKudr — APPROVED** [review]
Works well, tasks are created without any errors on the server.
Screenshot 2024-12-04 095322
Screenshot 2024-12-04 095413

**Review by iLLiCiTiT — APPROVED** [review]

File: services/shotgrid_common/utils.py [hunk 1]
```diff
@@ -119,15 +121,19 @@ def _sg_to_ay_dict(
     if sg_entity["type"] == "Task":
         ay_entity_type = "task"
         if not sg_entity["step"]:
-            raise ValueError(
-                f"Task {sg_entity} has no Pipeline Step assigned."
+            log.warning(
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:124]:
```suggestion
            log.debug(
```

**Pull request merged**